### PR TITLE
optimize stream deadline timers

### LINKED ISSUE #1543: use a single timer in the stream
Currently, we're using `time.After` in the `sendStream` and the `receiveStream`. This creates a lot of allocations which aren't garbage collected until the timer expires (which can be far in the future). It's better to use a single timer and reset it.
See https://gist.github.com/marten-seemann/33a80e3b84c0a38e5416b51f18426326 for a benchmark:
```
BenchmarkTimeAfter-12     	 5000000	       304 ns/op	     253 B/op	       3 allocs/op
BenchmarkTimerReset-12    	20000000	        89.3 ns/op	       0 B/op	       0 allocs/op
```

### PR DESCRIPTION
**Pull request by marten-seemann**
Fixes #1543. Thanks to @kubuxu for pointing this out.

We're currently calling `time.After` for every change of the deadline, as long as a `Read` or a `Write` call is blocking. This creates a new timer which is only GCed after the deadline expires, and consumes about 250 bytes of data each.
By using a single timer, we can avoid any allocations for setting a new deadline.

### CODE REVIEW

File: receive_stream.go [hunk 1]
```diff
@@ -272,13 +272,22 @@ func (s *receiveStream) onClose(offset protocol.ByteCount) {
...
+	s.deadline = t
 	s.mutex.Unlock()
-	// if the new deadline is before the currently set deadline, wake up Read()
-	if t.Before(oldDeadline) {
+	if s.deadline.IsZero() { // skip if there's no deadline to set
 		s.signalRead()
+		return nil
+	}
+	// Lazily initialize the deadline timer.
+	if s.deadlineTimer == nil {
```
Comment (lucas-clemente) [inline receive_stream.go]: Does this not need to be protected by the mutex? (same in the other commit)

Comment (marten-seemann) [reply receive_stream.go]: Good catch. I think we can just use a `defer` here, since the deadline calls won't be used that often.

Comment (Kubuxu) [reply receive_stream.go]: If you want it to be thread safe, then `sync.Once` might be the best option.

Comment (marten-seemann) [reply receive_stream.go]: It would, if we didn't have the `deadline`, which also needs to be thread-safe.

**Pull request merged**